pkg/retainer/txretainer/transaction_retainer.go

```diff
@@ -0,0 +1,315 @@
+			return signedTransactionMetadata.TransactionMetadata().EarliestIncludedAttachment().Slot()
+		}
+
+		e.Initialized.OnTrigger(func() {
+			// attaching the transaction failed for some reason => store the error
+			// HINT: we treat the transaction as unsigned here, because we don't know if it was signed or not.
+			// This should not be a problem, because the error reason will still be stored and visible to the user,
+			// it might get overwritten by an invalid attachment later, but that's fine.
+			e.Ledger.MemPool().OnAttachTransactionFailed(func(txID iotago.TransactionID, blockID iotago.BlockID, err error) {
+				r.updateTxMetadata(txID, &txMetadata{
+					TransactionID:          txID[:],
+					ValidSignature:         false,
+					EarliestAttachmentSlot: blockID.Slot(),
+					State:                  byte(api.TransactionStateFailed),
+					FailureReason:          byte(api.DetermineTransactionFailureReason(err)),
+					ErrorMsg:               getTxErrMsg(err),
+				})
+			}, asyncOpt)
+
+			// transaction was successfully attached to the tangle
+			e.Ledger.MemPool().OnTransactionAttached(func(transactionMetadata mempool.TransactionMetadata) {
+				e.LogTrace("TxRetainer.TransactionAttached", "tx", transactionMetadata.ID())
+
+				// store that we saw the transaction
+				txID := transactionMetadata.ID()
+				r.updateTxMetadata(txID, &txMetadata{
+					TransactionID:          txID[:],
+					ValidSignature:         false,
+					EarliestAttachmentSlot: transactionMetadata.EarliestIncludedAttachment().Slot(),
+					State:                  byte(api.TransactionStatePending),
+					FailureReason:          byte(api.TxFailureNone),
+					ErrorMsg:               nil,
+				})
+
+				/*
+					confirmed and finalized logic need to be done in the getter
+					with the slot of the earliest included attachment
+						first check if the earliest included attachment slot is finalized
+						if not, check if the attachment itself was confirmed
+					/*
+						// TransactionStateFailed indicates that the transaction has not been executed by the node due to a failure during processing.
+						TransactionStateFailed
+				*/
+
+				transactionMetadata.OnAccepted(func() {
+					e.LogTrace("TxRetainer.TransactionAccepted", "tx", transactionMetadata.ID())
+
+					// transaction was accepted
+					txID := transactionMetadata.ID()
+					r.updateTxMetadata(txID, &txMetadata{
+						TransactionID:          txID[:],
+						ValidSignature:         true, // if the transaction was accepted, the signature is valid
+						EarliestAttachmentSlot: transactionMetadata.EarliestIncludedAttachment().Slot(),
+						State:                  byte(api.TransactionStateAccepted),
+						FailureReason:          0,
+						ErrorMsg:               nil,
+					})
+				})
+
+				// a conflict that lost => double spent
+				transactionMetadata.OnRejected(func() {
+					e.LogTrace("TxRetainer.TransactionRejected", "tx", transactionMetadata.ID())
+
+					err := iotago.ErrInputAlreadySpent
```

Comment:
I think this should be the error:

```suggestion
					err := iotago.ErrTxConflictRejected
```